### PR TITLE
Chronos2 train

### LINKED ISSUE #272: Implement Chronos integration

### PR DESCRIPTION
**Pull request by RobotPsychologist**
## Chronos-2 Fine-Tuning Pipeline and Sweep Results

Adds a complete AutoGluon-backed Chronos-2 fine-tuning and evaluation
pipeline for nocturnal hypoglycemia probability forecasting, together with
results from a 16-config hyperparameter sweep, context window ablation, and
training convergence step sweep.

### New: Chronos-2 model module (`src/models/chronos2/`)
- `Chronos2Forecaster` backed by AutoGluon `TimeSeriesPredictor` (not a raw
  HuggingFace Trainer like TTM)
- Two-stage architecture: LoRA fine-tuning pipeline and midnight-anchored
  probabilistic inference
- Covariate support: IOB, COB, insulin_availability
- **Bug fix:** `use_cache=False` in `_predict_batch()` — required when
  evaluating multiple training checkpoints through AutoGluon snapshots.
  Snapshots share a symlinked `trainer.pkl`, so without this flag AutoGluon
  writes `cached_predictions.pkl` on the first call and returns it verbatim
  for every subsequent checkpoint. See `notes/chronos2/autogluon_snapshot_eval_gotcha.md`.

### New: 16-config hyperparameter sweep (`configs/models/chronos2/`)
- **Covariate ablations:** BG-only → +IOB → +IOB+COB → full features
- **LR ablations:** 1e-5 vs 5e-5
- **Context window ablations:** 64 / 128 / 256 / 512 steps (~5–43 h)
- All configs: 10k training steps, batch 256, checkpoint every 2k steps

### New: Scripts
- `scripts/training/`: local + SLURM fine-tuning scripts
- `scripts/experiments/`: sweep training, sweep eval, context ablation eval,
  step sweep eval, end-to-end orchestrators

### New: Experiment results
- **Context ablation** (`experiments/nocturnal_forecasting_ctx_ablation/`):
  96 runs comparing fine-tuned IOB configs at each context length against
  zero-shot baselines, using a fair fixed-episode holdout set.
- **Step sweep** (`experiments/nocturnal_forecasting_step_sweep/`):
  60 runs evaluating IOB configs 04/10/11/12 at training checkpoints
  2k–10k steps × Aleppo/Brown/Lynch datasets. Results are post-cache-fix.

### New: Analysis
- `scripts/analysis/plot_step_sweep.py`: 8×3 convergence grid, shared
  per-row y-axis, ideal coverage reference lines

### New: Tests
- `tests/models/test_chronos2.py` — integration tests
- `tests/models/gpu_smoke_test_chronos2_inference.py` — GPU smoke test

### New: Notes
- `notes/chronos2/sweep_plan.md` — sweep design
- `notes/chronos2/results_analysis.md` — full results with ctx-ablation
  conclusions (corrected post-cache-fix)
- `notes/chronos2/autogluon_snapshot_eval_gotcha.md` — caching trap in
  AutoGluon snapshot evaluation

### Key findings
Best config: **04 (IOB+IA, LR=5e-5, ctx=512)** — strongest on Aleppo and
Lynch across most metrics. Context 512 consistently beats shorter windows.
Step convergence mostly plateaus by 8k steps. Full tables in
`notes/chronos2/results_analysis.md`.

### CODE REVIEW

File: scripts/analysis/summarize_experiments.py [hunk 1]
```diff
@@ -34,9 +34,23 @@
...
+class _NocturnalCtxAblationSummarizer(NocturnalSummarizer):
+    """NocturnalSummarizer scoped to experiments/nocturnal_forecasting_ctx_ablation/."""
+
+    def __init__(self, experiments_root: str = "experiments") -> None:
+        # Bypass NocturnalSummarizer.__init__ to set a different experiment_type.
+        from src.experiments.base.experiment import ExperimentSummarizer
+
+        ExperimentSummarizer.__init__(
+            self, experiments_root, "nocturnal_forecasting_ctx_ablation"
+        )
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/analysis/summarize_experiments.py:47]: Good catch. _parse_run_dir was hardcoding the module-level _EXPERIMENT_TYPE constant instead of self.experiment_type, so the subclass's corrected type never made it into the output rows. Fixed in both NocturnalSummarizer and StandardForecastingSummarizer for consistency — the latter isn't subclassed today but had the same latent issue.

File: src/models/chronos2/config.py [hunk 2]
```diff
@@ -127,6 +146,29 @@ def get_autogluon_hyperparameters(self) -> Dict:
                 "cross_learning": False,
             }
         }
+        if self.fine_tune_batch_size is not None:
+            hp["Chronos2"]["fine_tune_batch_size"] = self.fine_tune_batch_size
+        if self.batch_size is not None:
+            hp["Chronos2"]["batch_size"] = self.batch_size
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/config.py:151]: Agreed. `fine_tune_batch_size` was being written into the dict unconditionally (even as `None`), which would override AutoGluon's default with an explicit `None` and made the guard below it dead code. Removed it from the initial dict so it follows the same conditional pattern as `batch_size` and `min_past`.

File: scripts/experiments/nocturnal_hypo_eval_ctx_ablation.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    # predict time as were present during training.
+    covariate_cols = args.covariate_cols
+    if (
+        covariate_cols is None
+        and hasattr(config, "covariate_cols")
+        and config.covariate_cols
+    ):
+        covariate_cols = config.covariate_cols
+        logger.info("Using covariates from model config: %s", covariate_cols)
+
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/experiments/nocturnal_hypo_eval_ctx_ablation.py]: Good catches. Removed the duplicate covariate resolution block — the single block before `resolved_config` now serves both the config dict and the `evaluate_nocturnal_forecasting()` call. Also replaced the `hasattr` + direct attribute access pattern with `getattr(config, "covariate_cols", None)`, which fixes the Pyright error that was already flagging those lines (the type checker couldn't see `covariate_cols` on the base `ModelConfig`).

File: scripts/experiments/run_sweep03_checkpoint_evals.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+fi
+
+N_STEPS=${#SNAPSHOT_DIRS[@]}
+N_DATASETS=$(echo "${DATASETS}" | wc -w)
+TOTAL=$(( N_STEPS * N_DATASETS ))
+
+echo "=== sweep-03 checkpoint evals ==="
+echo "  artifact dir : ${ARTIFACT_DIR}"
+echo "  snapshots    : ${N_STEPS}  ($(basename "$(dirname "${SNAPSHOT_DIRS[0]}")")  →  $(basename "$(dirname "${SNAPSHOT_DIRS[-1]}")"))"
+echo "  datasets     : ${DATASETS}"
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/experiments/run_sweep03_checkpoint_evals.sh]: Both fixed. The sort now uses awk to extract the numeric step value from the step_N directory basename as a dedicated sort key, so underscores anywhere in ARTIFACT_DIR can't shift fields. Swapped SNAPSHOT_DIRS[-1] for an explicit LAST_IDX=$(( N_STEPS - 1 )) — no behaviour change on modern Bash but avoids the Bash 3.2 incompatibility.

File: scripts/experiments/run_sweep03_checkpoint_evals.sh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    echo "ERROR: no snapshots/ dir found under ${ARTIFACT_DIR}" >&2
+    echo "       Has the 03_250k_checkpoints.yaml training run finished?" >&2
+    exit 1
+fi
+
+# Collect and sort snapshots by step number
+mapfile -t SNAPSHOT_DIRS < <(
+    find "${SNAPSHOTS_DIR}" -maxdepth 2 -name "model.pt" -type d \
+    | sort -t_ -k2 -n
+)
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/experiments/run_sweep03_checkpoint_evals.sh]: Both fixed. The sort now uses awk to extract the numeric step value from the step_N directory basename as a dedicated sort key, so underscores anywhere in ARTIFACT_DIR can't shift fields. Swapped SNAPSHOT_DIRS[-1] for an explicit LAST_IDX=$(( N_STEPS - 1 )) — no behaviour change on modern Bash but avoids the Bash 3.2 incompatibility.

File: scripts/visualization/plot_checkpoint_rmse_vs_steps.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        if not m or m.group(2) != dataset:
+            continue
+        results_file = d / RESULTS_FILE
+        if not results_file.exists():
+            continue
+        with open(results_file) as f:
+            data = json.load(f)
+        metric = data.get("overall_wql") or data.get("overall_rmse")
+        if metric is not None:
+            results.append((int(m.group(1)), metric))
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/visualization/plot_checkpoint_rmse_vs_steps.py:67]: Good catch — 0.0 is a valid WQL and would have silently fallen back to RMSE. Fixed with an explicit is not None check.

File: src/models/chronos2/config.py [hunk 7]
```diff
@@ -45,14 +52,19 @@ class Chronos2Config(ModelConfig):
...
     fine_tune_steps: int = 15000
     fine_tune_lr: float = 1e-5
-    fine_tune_batch_size: int = 32
+    # None = use AutoGluon defaults (fine_tune_batch_size=32, batch_size=256)
+    fine_tune_batch_size: Optional[int] = None
+    batch_size: Optional[int] = None
     time_limit: Optional[int] = None
+    # How often (in gradient steps) the HuggingFace Trainer logs loss/lr.
+    # 500 gives ~20 log lines over a 10k-step run. Set None to use HF default (500).
+    fine_tune_logging_steps: int = 500
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/config.py]: Agreed on both points. Changed to `Optional[int] = None` and wrapped the `logging_strategy`/`logging_steps` injection in a `is not None` guard, so runs without an explicit value get HF Trainer's default behaviour rather than a redundant override. The comment claiming `None` was valid was already misleading given the old `int` type — that's now consistent.

File: src/models/chronos2/config.py [hunk 8]
```diff
@@ -127,6 +146,29 @@ def get_autogluon_hyperparameters(self) -> Dict:
...
+                    "save_steps": self.checkpoint_save_steps,
+                    # Don't save optimizer/scheduler state — we only need weights.
+                    "save_only_model": True,
+                    # Override Chronos2 pipeline's hardcoded save_total_limit=1
+                    # so intermediate checkpoints are not deleted.
+                    "save_total_limit": None,
+                }
+            )
+        hp["Chronos2"]["fine_tune_trainer_kwargs"] = trainer_kwargs
         return hp
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/config.py]: Agreed. `fine_tune_trainer_kwargs` was always injected, including as {} when neither checkpoint_save_steps nor fine_tune_logging_steps was set — contradicting the test's assertion that it must be absent in that case. Fixed by guarding the injection with `if trainer_kwargs:`.

File: src/models/chronos2/model.py [hunk 9]
```diff
@@ -208,14 +210,202 @@ def _train_model(
...
+            os.makedirs(snapshot_model_pt, exist_ok=True)
+
+            # predictor_path must be absolute — _load_checkpoint resolves it against
+            # CWD, not against the JSON file location.
+            with open(
+                os.path.join(snapshot_model_pt, "chronos2_predictor.json"), "w"
+            ) as f:
+                json.dump(
+                    {"predictor_path": os.path.abspath(shadow_predictor)}, f, indent=2
+                )
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/model.py]: Good catch. Agreed — writing an absolute path made the snapshot artifacts non-relocatable, and the fallback to `model_dir` on a stale path was broken since `model.pt/` is not a valid `TimeSeriesPredictor` directory.

Fixed in three places:
- `_materialize_intermediate_checkpoints`: now writes `"../predictor"` (always stable given the `step_N/model.pt/` → `step_N/predictor/` layout)
- `_save_checkpoint`: now writes `os.path.relpath(predictor.path, output_dir)` instead of the absolute path
- `_load_checkpoint`: resolves relative `predictor_path` values against `os.path.dirname(ref_path)` before the existence check, so artifacts work after relocation. Absolute paths (existing artifacts on disk) continue to work unchanged via the `os.path.isabs` guard.

**Review by RobotPsychologist — COMMENTED** [review]
LGTM

File: src/experiments/base/experiment.py [hunk 10]
```diff
@@ -28,7 +28,7 @@
 
 log = logging.getLogger(__name__)
 
-# Run-directory naming convention: YYYY-MM-DD_HHMM_<dataset>_<mode>
+# Conventional run-directory naming: YYYY-MM-DD_HHMM_<dataset>_<mode>
 _RUN_DIR_RE = re.compile(r"^\d{4}-\d{2}-\d{2}_\d{4}_\S+$")
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/experiments/base/experiment.py]: Good catch. `nocturnal_hypo_eval.py` and `nocturnal_hypo_eval_ctx_ablation.py`
both use `%H%M%S` (6 digits), so every run they produce was hitting the
"Non-standard run directory name" debug log. Updated `_RUN_DIR_RE` to
`\d{4,6}` to accept both `HHMM` and `HHMMSS`, and updated the comment to
document both conventions. `sliding_window_eval.py` still uses `%H%M` and
continues to work unchanged.

File: src/models/chronos2/config.py [hunk 11]
```diff
@@ -86,6 +98,13 @@ class Chronos2Config(ModelConfig):
     # most windows naturally get full context regardless of this setting.
     min_past: int = 1
 
+    # Periodic checkpointing during fine-tuning.
+    # When set, the HuggingFace Trainer saves a checkpoint every N steps via
+    # fine_tune_trainer_kwargs. After training, _train_model materialises each
+    # checkpoint-N/ dir into a standalone eval-ready snapshot alongside the
+    # main output_dir so it can be passed directly to --checkpoint.
+    checkpoint_save_steps: Optional[int] = None
+
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/config.py:107]: I'll pass on this one. HF Trainer already raises a `ValueError` for
`save_steps <= 0`, so the failure is immediate and clearly attributed — there's
no silent undefined behavior to guard against. The `<= fine_tune_steps`
cross-field check would also prevent setting `checkpoint_save_steps` larger
than `fine_tune_steps` as a deliberate "no intermediate checkpoints" signal.
For internal research config I'd rather keep `__post_init__` lean and let the
downstream library surface misconfiguration.

File: src/models/chronos2/model.py [hunk 12]
```diff
@@ -208,14 +213,201 @@ def _train_model(
...
+            adapter_src = os.path.join(ckpt_dir, "adapter_model.safetensors")
+            if not os.path.exists(adapter_src):
+                info_print(f"  {ckpt_name}: no adapter_model.safetensors, skipping")
+                continue
+
+            snapshot_dir = os.path.join(snapshots_base, f"step_{step_num}")
+            if os.path.exists(snapshot_dir):
+                info_print(f"  step_{step_num}: already exists, skipping")
+                continue
+
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/model.py:300]: I think this is over-engineering for our use case. `_materialize_intermediate_checkpoints`
runs after the main predictor is fully saved, so a crash there doesn't affect
the primary artifact. If a partial snapshot causes an eval failure it fails
loudly with a missing-file error — not silently with corrupted results. And
recovery is just re-running training, which rebuilds everything from scratch.
The atomic-rename pattern adds meaningful complexity for a failure mode that
already surfaces clearly and has a one-liner fix. I'd rather keep it simple here.

File: src/models/chronos2/model.py [hunk 13]
```diff
@@ -375,6 +568,7 @@ def _autogluon_extract(self, data: pd.DataFrame, columns: list) -> np.ndarray:
             For multiple columns: shape (len(columns), n_episodes * forecast_length).
         """
         ts_data = self._prepare_autogluon_data(data)
+        assert self.predictor is not None  # only called in the fine-tuned path
         ag_predictions = self.predictor.predict(ts_data)
         multi = len(columns) > 1
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/models/chronos2/model.py:577]: Good catch. `_autogluon_extract()` had the same caching exposure as `_predict_batch()`.
Fixed with `use_cache=False` and the same explanatory comment. Option (a) is
the right call here — no added complexity, consistent behaviour across all
fine-tuned inference paths.
